dbms/src/Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.cpp

```diff
@@ -11,42 +11,51 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+#include <Common/setThreadName.h>
 #include <Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.h>
 #include <Storages/DeltaMerge/ReadThread/SegmentReader.h>
 #include <Storages/DeltaMerge/Segment.h>
 
 namespace DB::DM
 {
-SegmentReadTaskScheduler::SegmentReadTaskScheduler()
+SegmentReadTaskScheduler::SegmentReadTaskScheduler(bool run_sched_thread)
     : stop(false)
     , log(Logger::get())
 {
-    sched_thread = std::thread(&SegmentReadTaskScheduler::schedLoop, this);
+    if (likely(run_sched_thread))
+    {
+        sched_thread = std::thread(&SegmentReadTaskScheduler::schedLoop, this);
+    }
 }
 
 SegmentReadTaskScheduler::~SegmentReadTaskScheduler()
 {
     setStop();
-    sched_thread.join();
+    if (likely(sched_thread.joinable()))
+    {
+        sched_thread.join();
+    }
 }
 
-void SegmentReadTaskScheduler::add(const SegmentReadTaskPoolPtr & pool, const LoggerPtr & req_log)
+void SegmentReadTaskScheduler::add(const SegmentReadTaskPoolPtr & pool)
 {
     Stopwatch sw_add;
     // `add_lock` is only used in this function to make all threads calling `add` to execute serially.
     std::lock_guard add_lock(add_mtx);
+    add_waittings.fetch_add(1, std::memory_order_relaxed);
     // `lock` is used to protect data.
     std::lock_guard lock(mtx);
+    add_waittings.fetch_sub(1, std::memory_order_relaxed);
     Stopwatch sw_do_add;
-    read_pools.add(pool);
+    read_pools.emplace(pool->pool_id, pool);
```

Comment:
The `SegmentReadTaskScheduler::add` is public, it could be called by someone else. Wo we should assert `pool` to be not `nullptr`.